### PR TITLE
added tests for AqlValue in-memory optimization

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

This PR add tests for constructing AqlValues from unsigned integer values.
It fixes a bug when constructing AqlValues from uint64_t values that are larger than INT64_MAX.
It also unifies the handling of +0.0 and -0.0 in some cases and adds a few tests for this.

It does not add tests for constructing AqlValues from signed integers or from floating-point values. These tests can be added in separate PRs.

Parts of this (handling of +0.0 and -0.0) may need to be backported to 3.7 later.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
- [x] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/14049/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/14064/

File: arangod/Aql/Ast.cpp [hunk 1]
```diff
@@ -1124,6 +1124,10 @@ AstNode* Ast::createNodeValueInt(int64_t value) {
 
 /// @brief create an AST double value node
 AstNode* Ast::createNodeValueDouble(double value) {
+  if (value == -0.0) {
+    // unify -0.0 and +0.0 
+    value = 0.0;
+  }
```
Comment (mpoeter) [inline arangod/Aql/Ast.cpp:1130]: This is inconsistent with the `AqlValueHintDouble` constructor which maps NaN and infinity to null. Is this intentional? If we have to support doubles with NaN values, this complicates comparison if we want to comply to the IEEE 754 standard.
But if we do want to support NaNs, we should at least make sure that we don't have signaling NaNs to avoid invalid operation exceptions!

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+  AqlValueMemory memory;
+  void* p = reinterpret_cast<void*>(&memory);
+
+  // put uint64_t value into an AqlValue directly using the AqlValueHintUInt ctor.
+  // note: we are using placement new here, so that the AqlValue uses the
+  // poisoned memory region
+  new (p) AqlValue((AqlValueHintUInt(value)));
+  // although we are using placement new here, we don't need to call the
+  // destructor here, as the AqlValue with the payloads we use here won't
+  // do anything it is destructor
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:91]:
```suggestion
  // do anything in its destructor
```

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+  AqlValue& aqlValue = *reinterpret_cast<AqlValue*>(p);
+  runChecksForNumber(aqlValue, expected);
+
+  ASSERT_EQ(value != 0, aqlValue.toBoolean());
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:96]:
```suggestion
  EXPECT_EQ(value != 0, aqlValue.toBoolean());
```
I prefer to use EXPECT instead of ASSERT for assertions that are not a precondition for following code, because that way the test does not abort with the first failing EXPECT and we get more information. As far as I see all ASSERTs in this test file could be converted to EXPECTs.
I only use ASSERT in cases where we definitely want to abort, because otherwise we would run into issues. E.g.:
```
  ASSERT_NE(nullptr, x);
  EXPECT_EQ(42, *x);
```

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    AqlValue::AqlValueType::VPACK_INLINE_INT48, 0x2b, 0xff, 0xff, 0xff, 0x7f, UNINITIALIZED, UNINITIALIZED,
+    0xff, 0xff, 0xff, 0x7f, 0x00, 0x00, 0x00, 0x00
+  };
+  runChecksForUInt64(uint64_t(2147483647ULL), expected);
+}
+
+TEST(AqlValueMemoryLayoutTest, UnsignedLargerValues48Bit_2147483648) {
+  // 7f ff ff ff
+  uint8_t const expected[] = { 
+    AqlValue::AqlValueType::VPACK_INLINE_INT48, 0x2c, 0x00, 0x00, 0x00, 0x80, 0x00, UNINITIALIZED,
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:283]: Why is this value encoded with 5 byes? 4 would be enough..?

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    0x00, 0x00, 0x00, 0x80, 0x00, 0x00, 0x00, 0x00
+  };
+  runChecksForUInt64(uint64_t(2147483648ULL), expected);
+}
+
+TEST(AqlValueMemoryLayoutTest, UnsignedLargerValues48Bit_4294967294) {
+  // ff ff ff fe
+  uint8_t const expected[] = { 
+    AqlValue::AqlValueType::VPACK_INLINE_INT48, 0x2c, 0xfe, 0xff, 0xff, 0xff, 0x00, UNINITIALIZED,
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:292]: Why is this value encoded with 5 byes? 4 would be enough..?

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+    0xfe, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00
+  };
+  runChecksForUInt64(uint64_t(4294967294ULL), expected);
+}
+
+TEST(AqlValueMemoryLayoutTest, UnsignedLargerValues48Bit_4294967295) {
+  // ff ff ff ff
+  uint8_t const expected[] = { 
+    AqlValue::AqlValueType::VPACK_INLINE_INT48, 0x2c, 0xff, 0xff, 0xff, 0xff, 0x00, UNINITIALIZED,
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:301]: Why is this value encoded with 5 byes? 4 would be enough..?

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+    AqlValue::AqlValueType::VPACK_INLINE_INT48, 0x2c, 0xff, 0xff, 0xff, 0xff, 0x7f, UNINITIALIZED,
+    0xff, 0xff, 0xff, 0xff, 0x7f, 0x00, 0x00, 0x00
+  };
+  runChecksForUInt64(uint64_t(549755813887ULL), expected);
+}
+
+TEST(AqlValueMemoryLayoutTest, UnsignedLargerValues48Bit_549755813888) {
+  // 7f ff ff ff ff
+  uint8_t const expected[] = { 
+    AqlValue::AqlValueType::VPACK_INLINE_INT48, 0x2d, 0x00, 0x00, 0x00, 0x00, 0x80, 0x00,
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:328]: Why is this value encoded with 6 byes? 5 would be enough..?

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+  };
+  runChecksForUInt64(uint64_t(140737488355327ULL), expected);
+}
+
+// note: in all following tests, the value UNINITIALIZED (0xa5) has a special meaning.
+// if it used, it means that we don't compare the byte, because it is not initialized
+// in certain AqlValue configurations.
+// if you add tests, make sure that none of the test data includes 0xa5!
+TEST(AqlValueMemoryLayoutTest, UnsignedLargerValues64Bit_281474976710654) {
+  // 80 ff ff ff ff ff
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:357]:
```suggestion
  // ff ff ff ff ff fe
```

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+  uint8_t const expected[] = { 
+    AqlValue::AqlValueType::VPACK_INLINE_UINT64, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, 0x2f,
+    0xfe, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00
+  };
+  runChecksForUInt64(uint64_t(281474976710654ULL), expected);
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:362]: Why is this value encoded as a 64bit value? It only requires 48bits...

File: tests/Aql/AqlValueMemoryLayoutTest.cpp [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+}
+
+TEST(AqlValueMemoryLayoutTest, UnsignedLargerValues64Bit_281474976710655) {
+  // ff ff ff ff ff ff
+  uint8_t const expected[] = { 
+    AqlValue::AqlValueType::VPACK_INLINE_UINT64, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, UNINITIALIZED, 0x2f,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00
+  };
+  runChecksForUInt64(uint64_t(281474976710655ULL), expected);
```
Comment (mpoeter) [inline tests/Aql/AqlValueMemoryLayoutTest.cpp:371]: Why is this value encoded as a 64bit value? It only requires 48bits...

File: tests/Aql/AqlValueMemoryLayoutTest.cpp (hunk 3 above)
Comment (jsteemann) [reply tests/Aql/AqlValueMemoryLayoutTest.cpp]: Good idea, done that!

File: arangod/Aql/Ast.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/Ast.cpp:1134]: Fixed

File: tests/Aql/AqlValueMemoryLayoutTest.cpp (hunk 5 above)
Comment (jsteemann) [reply tests/Aql/AqlValueMemoryLayoutTest.cpp:292]: As discussed, this is necessary because the value is encoded as a signed integer right now.

File: tests/Aql/AqlValueMemoryLayoutTest.cpp (hunk 6 above)
Comment (jsteemann) [reply tests/Aql/AqlValueMemoryLayoutTest.cpp:301]: As discussed, this is necessary because the value is encoded as a signed integer right now.

File: tests/Aql/AqlValueMemoryLayoutTest.cpp (hunk 7 above)
Comment (jsteemann) [reply tests/Aql/AqlValueMemoryLayoutTest.cpp:328]: As discussed, this is necessary because the value is encoded as a signed integer right now.

File: tests/Aql/AqlValueMemoryLayoutTest.cpp (hunk 9 above)
Comment (jsteemann) [reply tests/Aql/AqlValueMemoryLayoutTest.cpp:362]: As discussed, this is necessary because the value is encoded as a signed integer right now.

File: tests/Aql/AqlValueMemoryLayoutTest.cpp (hunk 10 above)
Comment (jsteemann) [reply tests/Aql/AqlValueMemoryLayoutTest.cpp:371]: As discussed, this is necessary because the value is encoded as a signed integer right now.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/14075/

**Review by gnusi — APPROVED** [review]

File: tests/js/server/aql/aql-double-conv.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+      res = db._query(q2).toArray()[0];
+      assertNull(res);
+      
+      const q3 = `RETURN NOOPT(-1.0) / NOOPT(0.0)`;
+      res = db._query(q3).toArray()[0];
+      assertNull(res);
+      
+      const q4 = `RETURN NOOPT(-1.0) / NOOPT(-0.0)`;
+      res = db._query(q4).toArray()[0];
+      assertNull(res);
```
Comment (mpoeter) [inline tests/js/server/aql/aql-double-conv.js:66]: We should also add a test for `0.0 / 0.0` - in JavaScript this returns NaN.

**Review by mpoeter — APPROVED** [review]

File: tests/js/server/aql/aql-double-conv.js (hunk 11 above)
Comment (jsteemann) [reply tests/js/server/aql/aql-double-conv.js:66]: Good idea. Done!
